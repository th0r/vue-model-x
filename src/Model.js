import {Watcher} from './vue-internals';
import {defineWatchersProperty, markAsStatic} from './utils';

export class Model {

  constructor() {
    // Tells Vue that it doesn't need to convert all properties of this object to reactive because
    // we explicitly mark reactive properties with `@observable` and `@computed` decorators.
    // All other properties shouldn't be reactive.
    markAsStatic(this);
  }

  update(data) {
    Object.assign(this, data);
  }

  watch(expression, callback, options) {
    if (!this.hasOwnProperty('_watchers')) {
      defineWatchersProperty(this);
    }

    if (typeof callback === 'string') {
      const methodName = callback;

      callback = this[methodName];

      if (typeof callback !== 'function') {
        throw new TypeError(`VueModel#watch: model doesn't have "${methodName}" method`);
      }
    }

    const watcher = new Watcher(this, expression, callback, options);

    if (options && options.immediate) {
      callback.call(this, watcher.value);
    }

    return () => {
      watcher.teardown();
    };
  }

}
