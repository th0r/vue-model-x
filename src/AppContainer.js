import {storeMeta} from './storeDecorator';

export class AppContainer {
  constructor(state) {
    this.stores = Object.create(null);
    this.state = state || {};
  }

  getStore(StoreConstructor) {
    const meta = StoreConstructor[storeMeta];

    if (!meta) {
      throw new Error(
        `VueModelX: couldn't find store meta information for "${StoreConstructor.name}" class.\n` +
        'Did you forget to decorate it with `@store`?'
      );
    }

    let {name, deps} = meta;
    let store = this.stores[name];

    if (store) {
      return store;
    }

    store = new StoreConstructor();
    this.stores[name] = store;

    if (deps) {
      if (typeof deps === 'function') {
        deps = deps();
      }

      for (const propName of Object.keys(deps)) {
        store[propName] = this.getStore(deps[propName]);
      }
    }

    const state = this.state.hasOwnProperty(name) ? this.state[name] : undefined;

    if (store.init) {
      store.init(state);
    } else if (state) {
      throw new Error(
        `VueModelX: there is a serialized state for "${meta.name}" store but it doesn't implement "init" method`
      );
    }

    return store;
  }
}
