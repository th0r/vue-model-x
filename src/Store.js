import {Model} from './Model';

export class Store extends Model {}
