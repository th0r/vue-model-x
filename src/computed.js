import {Dep, Watcher} from './vue-internals';

const computedWatchers = Symbol('computedWatchers');
const noop = () => {};

export function computed(prototype, getterName, descriptor) {
  const getter = descriptor.get;

  if (!getter) {
    throw new TypeError(
      `${prototype.constructor.name}#${getterName} is not a getter so it can't be used as computed property`
    );
  }

  return {
    configurable: true,
    enumerable: true,
    get() {
      if (!Dep.target) {
        return getter.call(this);
      }

      if (!this.hasOwnProperty(computedWatchers)) {
        // Needed by Vue's `Watcher` constructor
        this._watchers = [];
        this[computedWatchers] = new Map();
      }

      if (!this[computedWatchers].has(prototype)) {
        this[computedWatchers].set(prototype, Object.create(null));
      }

      const watchers = this[computedWatchers].get(prototype);
      let watcher = watchers[getterName];

      if (!watcher) {
        // Lazily creating computed watcher
        watcher = watchers[getterName] = new Watcher(this, getter, noop, {lazy: true});
      }

      if (watcher.dirty) {
        watcher.evaluate();
      }

      watcher.depend();

      return watcher.value;
    }
  };
}
