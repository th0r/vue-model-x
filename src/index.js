export {observable} from './observable';
export {computed} from './computed';
export {VueModel} from './VueModel';
export {markAsStatic} from './utils';
