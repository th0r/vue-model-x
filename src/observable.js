import {Dep, defineReactive} from './vue-internals';

const observableValues = Symbol('observableValues');

export function observable(proto, name, descriptor) {
  Object.assign(descriptor, {
    configurable: true,
    writable: true
  });

  const reactiveDescriptor = {
    configurable: true,
    enumerable: true,

    get() {
      let value;

      if (this.hasOwnProperty(observableValues) && name in this[observableValues]) {
        value = this[observableValues][name];
      } else {
        value = descriptor.initializer();
        saveObservableValue(this, name, value);
      }

      if (Dep.target) {
        defineReactiveProperty(this, name, value, descriptor);
        // Should call newly created getter in order to set reactive dependency
        return this[name];
      } else {
        return value;
      }
    },

    set(value) {
      if (Dep.target) {
        defineReactiveProperty(this, name, value, descriptor);
      } else {
        if (!this.hasOwnProperty(name)) {
          // Defining reactive descriptor on the object itself in order to `Object.keys` to list it
          Object.defineProperty(this, name, reactiveDescriptor);
        }
        saveObservableValue(this, name, value);
      }
    }
  };

  return reactiveDescriptor;
}

function defineReactiveProperty(obj, name, value, originalDescriptor) {
  // Defining original property
  Object.defineProperty(obj, name, originalDescriptor);
  // Converting it to reactive
  defineReactive(obj, name, value);
}

function saveObservableValue(obj, name, value) {
  if (!obj.hasOwnProperty(observableValues)) {
    obj[observableValues] = Object.create(null);
  }

  obj[observableValues][name] = value;
}
