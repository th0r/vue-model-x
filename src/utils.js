export function defineWatchersProperty(obj) {
  Object.defineProperty(obj, '_watchers', {
    value: [],
    enumerable: false
  });
}
