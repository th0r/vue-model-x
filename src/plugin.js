import {AppContainer} from './AppContainer';

export function vueModelXPlugin(Vue, opts = {}) {
  let {appContainer} = opts;
  const appContainerProp = opts.injectProperty || '$app';

  if (!appContainer) {
    appContainer = new AppContainer(opts.state);
  }

  Vue.mixin({

    beforeCreate() {
      const {stores} = this.$options;

      this[appContainerProp] = appContainer;

      if (stores) {
        for (const propName of Object.keys(stores)) {
          this[propName] = appContainer.getStore(stores[propName]);
        }
      }
    }

  });
}
