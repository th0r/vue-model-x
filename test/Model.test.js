import Vue from 'vue';
import {mount} from '@vue/test-utils';
import UserComponent from './components/UserComponent';
import {User} from './models/User';
import {defineWatchersProperty} from '../src/utils';

let user;
let comp;
const createUser = () => user = new User({
  age: 30,
  address: {
    city: 'Domodedovo'
  },
  roles: ['admin']
});
const createComponent = (user = createUser()) => {
  comp = mount(UserComponent, {propsData: {user}, sync: false});
};

describe('Model', function () {
  let cb;

  beforeEach(() => {
    createUser();
    cb = jest.fn();
  });

  describe('#watch', function () {

    it('should invoke callback if watch expression changes', async function () {
      user.watch(() => [user.firstName, user.isAdult], cb);
      expect(cb).not.toBeCalled();
      user.firstName = 'Chuck';
      expect(cb).not.toBeCalled();
      await Vue.nextTick();
      expect(cb).toBeCalledWith(['Chuck', true], ['', true]);
      user.age = 17;
      expect(cb).toHaveBeenCalledTimes(1);
      await Vue.nextTick();
      expect(cb).toBeCalledWith(['Chuck', false], ['Chuck', true]);
      user.firstName = 'Alisa';
      user.age = 18;
      expect(cb).toHaveBeenCalledTimes(2);
      await Vue.nextTick();
      expect(cb).toHaveBeenCalledTimes(3);
      expect(cb).toBeCalledWith(['Alisa', true], ['Chuck', false]);
    });

    it('should allow simple path as watch expression', async function () {
      user.watch('address.city', cb);
      user.address.city = 'Moscow';
      await Vue.nextTick();
      expect(cb).toBeCalledWith('Moscow', 'Domodedovo');
    });

    it('should allow method name as callback', async function () {
      user.callbackMethod = cb;
      user.watch('age', 'callbackMethod');
      user.age = 18;
      await Vue.nextTick();
      expect(user.callbackMethod).toBeCalledWith(18, 30);
    });

    it("should throw if callback is a method name and this method doesn't exist", function () {
      expect(() => user.watch('age', 'nonExistingMethod')).toThrowErrorMatchingSnapshot();
    });

    it('should return function that stops watcher', async function () {
      const stopWatcher = user.watch('age', cb);
      user.age = 18;
      await Vue.nextTick();
      expect(cb).toHaveBeenCalledTimes(1);
      stopWatcher();
      user.age = 19;
      await Vue.nextTick();
      expect(cb).toHaveBeenCalledTimes(1);
    });

    it('should work if model already has `_watchers` property', async function () {
      expect(user._watchers).not.toBeDefined();
      defineWatchersProperty(user);
      expect(user._watchers).toBeDefined();
      user.watch('age', cb);
      user.age = 18;
      await Vue.nextTick();
      expect(cb).toBeCalledWith(18, 30);
    });

    describe('options', function () {
      describe('immediate', function () {
        it('should immediately invoke callback', async function () {
          user.watch('age', cb, {immediate: true});
          expect(cb).toBeCalledWith(30);
          expect(cb).toHaveBeenCalledTimes(1);
          user.age = 18;
          await Vue.nextTick();
          expect(cb).toBeCalledWith(18, 30);
          expect(cb).toHaveBeenCalledTimes(2);
        });

        it('should immediately invoke callback as method name', async function () {
          user.callbackMethod = cb;
          user.watch('age', 'callbackMethod', {immediate: true});
          expect(user.callbackMethod).toBeCalledWith(30);
        });
      });

      describe('deep', function () {
        it('should invoke callback when deep property changes', async function () {
          user.watch('address', cb, {deep: true});
          user.address.city = 'Moscow';
          await Vue.nextTick();
          expect(cb).toBeCalledWith(user.address, user.address);
        });

        it('should be `false` by default', async function () {
          user.watch('address', cb);
          user.address.city = 'Moscow';
          await Vue.nextTick();
          expect(cb).not.toBeCalled();
        });
      });

      describe('sync', function () {
        it('should invoke callback immediately after change', async function () {
          user.watch('age', cb, {sync: true});
          expect(cb).not.toBeCalled();
          user.age = 18;
          expect(cb).toBeCalledWith(18, 30);
          await Vue.nextTick();
          expect(cb).toHaveBeenCalledTimes(1);
        });

        it('should be `false` by default', async function () {
          user.watch('age', cb);
          expect(cb).not.toBeCalled();
          user.age = 18;
          expect(cb).not.toBeCalled();
          await Vue.nextTick();
          expect(cb).toHaveBeenCalledTimes(1);
        });
      });
    });

  });

  it('component should not react to changes in non-reactive model properties', async () => {
    createComponent();
    const defaultRoles = user.roles.join(', ');
    expect(comp.find('#roles').text()).toEqual(defaultRoles);
    user.roles.push('admin');
    await Vue.nextTick();
    expect(comp.find('#roles').text()).toEqual(defaultRoles);
    user.roles = ['foo'];
    await Vue.nextTick();
    expect(comp.find('#roles').text()).toEqual(defaultRoles);
  });

});
