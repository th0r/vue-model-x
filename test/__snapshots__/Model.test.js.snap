// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model #watch should throw if callback is a method name and this method doesn't exist 1`] = `"VueModel#watch: model doesn't have \\"nonExistingMethod\\" method"`;
